src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+
+            var root = await currentDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, generator);
+            var syntaxFacts = currentDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+            var insertionIndex = GetInsertionIndex(compilation, methodSymbol, syntaxFacts, containingMethod);
+            var newMethodIdentifier = methodSymbol.Name + "_" + parameterName;
+
+            if (trampoline)
+            {
+                foreach (var invocationExpression in invocations)
+                {
+                    editor.ReplaceNode(invocationExpression, (currentInvocation, _) =>
+                    {
+                        return GenerateNewInvocationExpressionForTrampoline(syntaxFacts, editor, currentInvocation,
+                            invocationExpression, newMethodIdentifier, insertionIndex);
+                    });
+                }
+            }
+
+            // If you are at the original document, then also introduce the new method and introduce the parameter.
+            if (currentDocument.Id == originalDocument.Id)
+            {
+                if (trampoline)
+                {
+                    await GenerateNewMethodAsync(originalDocument, expression, methodSymbol, containingMethod, newMethodIdentifier,
+                        editor, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    GenerateNewMethodOverload(expression, methodSymbol, containingMethod, editor);
+                }
+
+                await UpdateExpressionInOriginalFunctionAsync(originalDocument, expression, containingMethod,
+                    parameterName, editor, allOccurrences, cancellationToken).ConfigureAwait(false);
+                var semanticModel = await originalDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var parameterType = semanticModel.GetTypeInfo(expression, cancellationToken).ConvertedType ?? semanticModel.Compilation.ObjectType;
+                var refKind = syntaxFacts.GetRefKindOfArgument(expression);
+                var parameter = generator.ParameterDeclaration(name: parameterName, type: generator.TypeExpression(parameterType), refKind: refKind);
+                editor.InsertParameter(containingMethod, insertionIndex, parameter);
+            }
+
+            return modifiedSolution.WithDocumentSyntaxRoot(originalDocument.Id, editor.GetChangedRoot());
+
+            // Adds an argument which is an invocation of the newly created method to the callsites
+            // of the method invocations where a parameter was added.
+            // Example:
+            // public void M(int x, int y)
+            // {
+            //     int f = x * y; // highlight this expression
+            // }
+            // 
+            // public void M1()
```

Comment:
fyi, your names are *Very* confusing.  please name this distinctly differently so that in the 'after' it isn't confusing with the other method with a very similar name.

Reply:
changed